### PR TITLE
feat: rework realtime subscription hooks

### PR DESCRIPTION
**Pull request by Ehesp**
Fixes https://github.com/invertase/react-query-firebase/issues/25

This PR reworks hooks that depend on realtime subscriptions.  Currently, if a hook declares a realtime subscription (via the `subscribe: true` option or by default (e.g. auth hooks)), a `staleTime` of `Infinity` is being set by default. If other hooks with the same `QueryKey` are created, the existing subscription will be used (which is expected).

The issue is, "Infinity" data is never considered stale for the remainder of the applications lifecycle. So assuming there is always an active subscription then this doesn't cause any issues (since the subscription always updates the cache). However, if all active hooks are unmounted, and remounted, the actual `queryFn` won't be executed again (since it'll just return the data within the query cache). However since all hooks have actually unmounted, there is no active subscription thus the data will never be updated again.

This PR basically reworks the effected hooks by;

1. Tracking the query key "mounts"
2. If the is existing query keys, then it'll simple return the cached data
3. If there is no existing query keys, it'll create a new subscription & store that reference
4. Each time a hook unmounts, the query key is subtracted from the active counter. If it's zero (no more hooks active), it'll unsubscribe.

### TODOs

- [x] Auth 
- [x] Firestore
- [ ] RTDB
- [ ] Add tests to test these conditions
- [ ] Extract / publish the `utils` package (since it causes weird building if it's directly import via rollup).
- [ ] Ensure all files have copyright headers

### CODE REVIEW

**Comment by atanaskanchev** [comment]
Hello, is there anything we can help to push this PR over the line?

**Comment by estebanrao** [comment]
Happy to help as well, this fix is crucial for an implementation in prod :)

**Comment by atanaskanchev** [comment]
> Happy to help as well, this fix is crucial for an implementation in prod :)

I'm also stuck and can't go to production without this fix. I've had great hopes for this lib as it was used by Google showcasing Firebase 9, sadly it seems the repo is abandoned, I'll start looking for alternatives.

**Comment by hchor** [comment]
Has anyone else tested this? It looks like the owner says it should work but is concerned about edge cases.

**Comment by Ehesp** [comment]
I'm going to try and get some resource from the company assigned to this soon, apologies it's been really busy.

**Comment by hchor** [comment]
> I'm going to try and get some resource from the company assigned to this soon, apologies it's been really busy.

This can't be stated enough but the community greatly appreciates your work!

**Comment by cabljac** [comment]
(just an update)

Currently working on this, experiencing some strange behaviour with the hooks in a test, `onSuccess` is called more times than i'd expect when rerendering.

**Comment by cabljac** [comment]
OK actually made some progress on this, curious what people think. I'm using a library called `react-query-subscriptions` which seems to work fairly well. Interested in any opinions here. A work in progress example:


```typescript
import { QueryKey } from "react-query";
import { Auth, IdTokenResult, AuthError } from "firebase/auth";
import { Observable } from "rxjs";
import {
  useSubscription,
  UseSubscriptionOptions,
} from "react-query-subscription";
import { UseSubscriptionResult } from "react-query-subscription/types/use-subscription";

export function idTokenFromAuth(
  auth: Auth,
  options?: {
    forceRefresh?: boolean;
  }
): Observable<IdTokenResult | null> {
  return new Observable<IdTokenResult | null>(function subscribe(subscriber) {
    const unsubscribe = auth.onIdTokenChanged(async (user) => {
      let token: IdTokenResult | null = null;

      if (user) {
        token = await user.getIdTokenResult(options?.forceRefresh);
      }

      subscriber.next(token);
    });
    subscriber.add(unsubscribe);
  });
}

export function useAuthIdToken<R = IdTokenResult | null>(
  key: QueryKey,
  auth: Auth,
  options?: {
    forceRefresh?: boolean;
  },
  useSubscriptionOptions?: Omit<
    UseSubscriptionOptions<IdTokenResult | null, AuthError, R>,
    "queryFn"
  >
): UseSubscriptionResult<R | AuthError> {
  return useSubscription(key, () => idTokenFromAuth(auth, options), {
    ...useSubscriptionOptions,
  });
}
```

**Comment by cabljac** [comment]
tests still need fixing etc, and will rewrite the rest of the subscriptions like this.

**Comment by estebanrao** [comment]
Is adding `react-query-subscription` really necessary? A dependency like this can cause issues in the future is no one maintains that package

**Comment by cabljac** [comment]
> Is adding `react-query-subscription` really necessary? A dependency like this can cause issues in the future is no one maintains that package

This is a fair point to be honest, it just seemed like they handled subscriptions quite nicely. @Ehesp What do you think?

**Comment by cabljac** [comment]
Yeah `react-query-subscription` seems to contain a few extra pieces we don't need. I think the solution to this bug is there, so i'll try and do something similar.

**Comment by estebanrao** [comment]
@cabljac great! let us know when we can jump in and test! :)

**Comment by cabljac** [comment]
hm seems a bit trickier to implement than initially hoped.

**Comment by estebanrao** [comment]
Even replicating the parts you need from `react-query-subscription`?

**Comment by cabljac** [comment]
Made some headway! essentially tweaking @Ehesp's solution a bit, but taking some ideas from `react-query-subscription` too. It seems to be passing tests.

**Comment by cabljac** [comment]
I think I've refactored all the subscription hooks now:

- useAuthUser
- useAuthIdToken
- useFirestoreDocument
- useFirestoreDocumentData
- useFirestoreQuery
- useFirestoreQueryData
- useDatabaseValue

If anyone notices there's one missing in this list let me know :). I've also added tests for the issue, which seem to pass. 
I need to ensure all files have copyright headers, and give things another look/ add a couple of code comments to the new useSubscription utility hook.

**Comment by atanaskanchev** [comment]
@cabljac amazing work, all seems to be in order. Should we as part of this PR bump the react dependencies to 18? I've used it with react 18 for some time now and haven't faced any issues

File: packages/database/src/useDatabaseSnapshot.ts [hunk 1]
```diff
@@ -1,61 +1,36 @@
...
+        return callback(snapshot);
       });
     },
-  });
+    [ref]
+  );
+
+  return useSubscription<DataSnapshot, Error, R>(
+    queryKey,
+    ["useFirestoreDatabase", ref.key],
```
Comment (yunseop-dev) [inline packages/database/src/useDatabaseSnapshot.ts:28]: useFirestoreDatabase does not seem appropriate.

File: packages/utils/src/useSubscription.ts [hunk 2]
```diff
@@ -0,0 +1,121 @@
...
+        observerCount[subscriptionHash] -= 1;
+        cleanupSubscription(subscriptionHash);
+      };
+    }
+  }, []);
+
+  const hashFn = options?.queryKeyHashFn || hashQueryKey;
+  const subscriptionHash = hashFn(subscriptionKey);
+
+  if (!options?.onlyOnce) {
```
Comment (yunseop-dev) [inline packages/utils/src/useSubscription.ts:53]: realtime database's subscription flag value is not `onlyOnce` but `subscribe`.

It seems that refactoring is needed.

File: packages/database/src/useDatabaseSnapshot.ts [hunk 3]
```diff
@@ -80,49 +55,29 @@ export type UseDatabaseValueOptions = {
...
+  );
+
+  return useSubscription<T, Error, R>(
+    queryKey,
+    ["useFirestoreDatabase", ref.key],
+    subscribeFn,
+    useQueryOptions,
+    !isSubscription,
+    async () => parseDataSnapshot(await get(ref), !!options?.toArray)
+  );
```
Comment (yunseop-dev) [inline packages/database/src/useDatabaseSnapshot.ts:82]:
```suggestion
return useSubscription<T, Error, R>(
    queryKey,
    ['useDatabaseValue', queryKey],
    subscribeFn,
    {
      ...useQueryOptions,
      onlyOnce: !isSubscription,
      fetchFn: async () => parseDataSnapshot(await get(ref), !!options?.toArray),
    },
  );
```
1. `useSubscription` has 4 args. The code update seems to be missing.
2. The reason for changing from 'ref.key' to 'queryKey' is that I found that a bug occurs when 'ref' is used as a query. When I changed it to 'queryKey', I checked that it works well.

File: packages/database/src/useDatabaseSnapshot.ts [hunk 4]
```diff
@@ -1,61 +1,36 @@
...
+
+  return useSubscription<DataSnapshot, Error, R>(
+    queryKey,
+    ["useFirestoreDatabase", ref.key],
+    subscribeFn,
+    useQueryOptions,
+    !isSubscription,
+    async () => get(ref)
+  );
```
Comment (yunseop-dev) [inline packages/database/src/useDatabaseSnapshot.ts:33]:
```suggestion
return useSubscription<DataSnapshot, Error, R>(
    queryKey,
    ['useDatabaseSnapshot', queryKey],
    subscribeFn,
    {
      ...useQueryOptions,
      onlyOnce: !isSubscription,
      fetchFn: async () => get(ref),
    },
  );
```

**Comment by cabljac** [comment]
OK so last bit of this is to add at least a single test for the error case, but that requires setting up firestore.rules and changing test setup a bit.

**Comment by cabljac** [comment]
a couple of tests pass on their own, but fail when run in series, and I can't figure out why. I skipped them and merged this. If people get a chance and are able to test it out, that would be greatly appreciated :D

**Pull request merged**